api/oc_rep_encode_crc.c

```diff
@@ -0,0 +1,156 @@
+{
+  uint64_t crc = 0;
+  size_t size = 0;
+  if (encoder->data.ptr != NULL) {
+    if ((encoder->flags & CrcFlagHasPayload) != 0) {
+      encoder->end = encoder->data.ptr + sizeof(crc);
+    }
+    size = encoder->end - encoder->data.ptr;
+  }
+
+  if (size < sizeof(crc)) {
+    encoder->end = NULL;
+    encoder->data.bytes_needed = (ptrdiff_t)(sizeof(crc) - size);
+    return CborErrorOutOfMemory;
+  }
+
+  if ((encoder->flags & CrcFlagHasPayload) != 0) {
+    memcpy(&crc, encoder->data.ptr, sizeof(crc));
+  }
+
+  crc = oc_crc64(crc, data, len);
+  memcpy(encoder->data.ptr, &crc, sizeof(crc));
+  encoder->flags |= CrcFlagHasPayload;
+  return CborNoError;
+}
+
+static CborError
+rep_crc_encode_null(CborEncoder *encoder)
+{
+  uint8_t crc_null[] = { OC_CRC_REP_NULL };
+  return rep_crc_append_to_buffer(encoder, crc_null, OC_ARRAY_SIZE(crc_null));
+}
+
+static CborError
+rep_crc_encode_bool(CborEncoder *encoder, bool value)
+{
+  uint8_t crc_bool[] = { value ? OC_CRC_REP_TRUE : OC_CRC_REP_FALSE };
+  return rep_crc_append_to_buffer(encoder, crc_bool, OC_ARRAY_SIZE(crc_bool));
+}
+
+static CborError
+rep_crc_encode_int(CborEncoder *encoder, int64_t value)
+{
+  return rep_crc_append_to_buffer(encoder, (const uint8_t *)&value,
+                                  sizeof(value));
+}
+
+static CborError
+rep_crc_encode_floating_point(CborEncoder *encoder, CborType fpType,
+                              const void *value)
+{
+  uint8_t size = sizeof(double);
+  if (fpType == CborHalfFloatType) {
+    size = sizeof(uint16_t);
+  } else if (fpType == CborFloatType) {
+    size = sizeof(float);
+  }
+  return rep_crc_append_to_buffer(encoder, (const uint8_t *)value, size);
+}
+
+static CborError
+rep_crc_encode_double(CborEncoder *encoder, double value)
+{
+  return rep_crc_append_to_buffer(encoder, (const uint8_t *)&value,
+                                  sizeof(value));
+}
+
+static CborError
+rep_crc_encode_uint(CborEncoder *encoder, uint64_t value)
+{
+  return rep_crc_append_to_buffer(encoder, (const uint8_t *)&value,
+                                  sizeof(value));
+}
+
+static CborError
+rep_crc_encode_byte_string(CborEncoder *encoder, const uint8_t *string,
+                           size_t length)
+{
+  return rep_crc_append_to_buffer(encoder, string, length);
+}
+
+oc_rep_encoder_t
+oc_rep_crc_encoder(void)
+{
+  return (oc_rep_encoder_t){
+    .type = OC_REP_CRC_ENCODER,
+
+    .get_buffer_size = &rep_crc_get_buffer_size,
+    .get_extra_bytes_needed = &rep_crc_get_extra_bytes_needed,
+
+    .encode_null = &rep_crc_encode_null,
+    .encode_boolean = &rep_crc_encode_bool,
+    .encode_int = &rep_crc_encode_int,
+    .encode_uint = &rep_crc_encode_uint,
+    .encode_floating_point = &rep_crc_encode_floating_point,
+    .encode_double = rep_crc_encode_double,
+    .encode_byte_string = &rep_crc_encode_byte_string,
+    .encode_text_string = NULL,
+    .create_array = NULL,
+    .create_map = NULL,
+    .close_container = NULL,
```

Comment:
What happens when they are not set ? It seems they throw segfault. eg:
```
CborError err = g_rep_encoder.close_container(encoder, containerEncoder);
```
Could you implement them similar as `encode_null` ?

Reply:
Yes, I will this PR is not finished, I just wanted to get some results from SC.